[SITE-5259] Add PHPUnit test coverage and Multidev test execution to CI

rkunjappan opened the pull request:
## Summary

- Add GitHub Actions CI pipeline that runs PHPUnit tests against Pantheon Multidev environments for both Drupal 10 and Drupal 11
- Add new unit tests for `GraphQL` and `PantheonContentPublisherConverter` classes
- Add kernel test for entity reference field creation targeting `pantheon_document.`
- Fix existing kernel tests for reliability and correctness

## Changes

### CI Pipeline (`.github/workflows/`, `.github/scripts/`)
- Extend `pull_request.yml` with a `deploy-multidev` job that creates Multidev environments on Drupal 10 and 11 fixture sites, installs the module, and runs Unit + Kernel test suites remotely via Terminus
- Add `create-multidev.sh` script: creates a Multidev, installs the module via Composer from the PR branch, and enables it with Drush
- Add `run-tests.sh` script: orchestrates remote PHPUnit execution with fail-fast (Unit then Kernel), redacts DB credentials from logs, and writes results to GitHub Step Summary
- Add Composer dependency caching, pin action SHAs, and fix concurrency group naming to prevent mirror/PR collisions
- Add `push`, `repository_dispatch`, and `workflow_dispatch` triggers with optional tmate debugging

### New Tests
- `tests/src/Unit/GraphQLTest.php` — Tests `getArticle`, `getArticleIds`, `getArticles`, `getMetadata` with publishing level and pagination variants, plus error handling for HTTP failures and missing response data
- `tests/src/Unit/PantheonContentPublisherConverterTest.php` — Tests data conversion, date format handling (ms timestamps, `msSinceEpoch` arrays), field mapping with dot-notation methods, metadata override priority, and edge cases (missing fields, unpublished status)
- `tests/src/Kernel/PantheonDocumentSelectionTest.php` — Tests entity reference field creation and loading for `pantheon_document` target type

### Test Infrastructure
- Add `phpunit.xml.dist` and `tests/bootstrap.php` for standalone unit test execution from the module root
- Add `autoload` / `autoload-dev` PSR-4 entries to `composer.json`
- Add `phpunit/phpunit` as a dev dependency

### Test Fixes
- Normalize GraphQL query whitespace in mock matching to prevent false mismatches
- Process queue items and re-index Search API after webhook execution in tests
- Reset entity cache before webhook processing to mirror production behavior
- Use valid JSON content in draft preview tests for `PantheonTagsFormatter` compatibility
- Add `#[RunTestsInSeparateProcesses]` to kernel tests to prevent cross-test state leakage
- Register `initial preprocess` callbacks in theme hook definitions
- Add `@testdox` annotations for readable test output

### Bug Fixes
- Wrap `GraphQL::request()` HTTP call in try/catch, throwing `GraphQLException` with the query name for better error diagnostics
- Fix `composer.json` license identifier to `GPL-2.0-or-later`


## Deprecation Warnings in D11 Kernel Tests (ignored)

The kernel tests on Drupal 11.3.x produce 46 deprecation warnings. These are **not test failures** — tests pass with `SYMFONY_DEPRECATIONS_HELPER=weak`. Here's a breakdown by source:

### Drupal Core (not actionable by this module)
- Twig/Symfony interface return type hints (`getFilters()`, `getFunctions()`, `getNodeVisitors()`, `getTokenParsers()`, `findTemplate()`, `exists()`) — Drupal core classes haven't added native return types yet
- `Twig\Environment::getTemplateClass()` internal method extended by `TwigEnvironment`
- `cache.backend.memory` and `cache.static` services deprecated in 11.3.0
- `Select::hasAllTags()`/`hasAnyTag()` and `QueryBase` signature changes
- `EntityListBuilder::getOperations()` / `ConfigEntityListBuilder::getDefaultOperations()` new `$cacheability` argument
- Symfony `ValidKeysConstraint` constructor deprecation

### Contrib Modules (not actionable by this module)
- **search_api**: `TaskStorageSchema` deprecated, `search_api_requirements` / `search_api_db_defaults_requirements` missing `#[LegacyRequirementsHook]`
- **key**: `@EntityType` annotation for `key` / `key_config_override`, `KeyPluginManager` missing attribute discovery

### This Module (tracked for future work)
- **`@EntityType` annotations**: `pantheon_document`, `pantheon_document_collection`, `pantheon_smart_component`, `pantheon_smart_instance` use annotations instead of PHP 8 attributes — migration to `#[EntityType]` should be done before  adding compatibility for Drupal 13
- **`#[RunTestsInSeparateProcesses]`**: Required on kernel test classes starting in 11.3.0 — already added to most test classes in this PR

### Why `SYMFONY_DEPRECATIONS_HELPER=weak`
The CI pipeline uses `weak` mode because the majority of deprecations originate in Drupal core and contrib dependencies that we cannot fix. Strict mode would cause false failures unrelated to this module's code.

rkunjappan committed:
Add phpunit tests integration with GitHub Actions

rkunjappan committed:
Restrict push trigger to 1.0.x branch to prevent duplicate CI runs

rkunjappan committed:
Fix kernel tests

rkunjappan committed:
Fix kernel tests for smart component and draft content

rkunjappan committed:
debugging smart component test error

rkunjappan committed:
Fix kernel test

rkunjappan committed:
More fixes for kernel test

rkunjappan committed:
Fix X-Frame-Options removal by using request attributes

rkunjappan committed:
Revert "Fix X-Frame-Options removal by using request attributes"

This reverts commit aa219ca743593765f74f15acc39a759c1a972227.

rkunjappan committed:
Add mock for REALTIME publishingLevel GraphQL query

rkunjappan committed:
fix code linting

rkunjappan committed:
Add tests for PantheonContentPublisherConverter and GraphQL

rkunjappan committed:
Add test for entity reference field creation

rkunjappan committed:
Fix unit tests

rkunjappan committed:
Fix issues with unit tests

rkunjappan committed:
Fix PantheonContentPublisherConverterTest

rkunjappan committed:
Fix tests

rkunjappan committed:
Fix CI workflow permissions and simplify PHP compatibility check

Add explicit permissions block to mirror.yml to satisfy GitHub
Advanced Security requirements. Consolidate phpcompatibility job
from a 3-version matrix to a single run using version range 8.2-.

rkunjappan committed:
Add explicit permissions to deploy-multidev job

Satisfy GitHub Advanced Security scanner by adding job-level
permissions block in addition to the workflow-level one.

rkunjappan committed:
Add explicit permissions to all jobs in pull_request workflow

Add job-level permissions blocks to linting and phpcompatibility
jobs to satisfy GitHub Advanced Security scanner.

rkunjappan committed:
Clarify comment for multidev existence check

Updated comment to clarify multidev existence check.

rkunjappan commented:
## Deprecation Warnings in D11 Kernel Tests (46 total)

The kernel tests on Drupal 11.3.x produce 46 deprecation warnings. These are **not test failures** — tests pass with `SYMFONY_DEPRECATIONS_HELPER=weak`. Here's a breakdown by source:

### Drupal Core (not actionable by this module)
- Twig/Symfony interface return type hints (`getFilters()`, `getFunctions()`, `getNodeVisitors()`, `getTokenParsers()`, `findTemplate()`, `exists()`) — Drupal core classes haven't added native return types yet
- `Twig\Environment::getTemplateClass()` internal method extended by `TwigEnvironment`
- `cache.backend.memory` and `cache.static` services deprecated in 11.3.0
- `Select::hasAllTags()`/`hasAnyTag()` and `QueryBase` signature changes
- `EntityListBuilder::getOperations()` / `ConfigEntityListBuilder::getDefaultOperations()` new `$cacheability` argument
- Symfony `ValidKeysConstraint` constructor deprecation

### Contrib Modules (not actionable by this module)
- **search_api**: `TaskStorageSchema` deprecated, `search_api_requirements` / `search_api_db_defaults_requirements` missing `#[LegacyRequirementsHook]`
- **key**: `@EntityType` annotation for `key` / `key_config_override`, `KeyPluginManager` missing attribute discovery

### This Module (tracked for future work)
- **`@EntityType` annotations**: `pantheon_document`, `pantheon_document_collection`, `pantheon_smart_component`, `pantheon_smart_instance` use annotations instead of PHP 8 attributes — migration to `#[EntityType]` should be done before  adding compatibility for Drupal 13
- **`#[RunTestsInSeparateProcesses]`**: Required on kernel test classes starting in 11.3.0 — already added to most test classes in this PR

### Why `SYMFONY_DEPRECATIONS_HELPER=weak`
The CI pipeline uses `weak` mode because the majority of deprecations originate in Drupal core and contrib dependencies that we cannot fix. Strict mode would cause false failures unrelated to this module's code.

tundeafolabi commented on .github/workflows/pull_request.yml:
```diff
@@ -3,31 +3,147 @@ on:
+        run: |
+          CURRENT_ENV="${MULTIDEV_ENV:-}"
+          PREFIX="ci-d${{ matrix.drupal-version }}-"
+          for env in $(terminus multidev:list "$TERMINUS_SITE" --format=list 2>/dev/null); do
```
should we use a script file for this logic ?

rkunjappan replied:
@tundeafolabi I moved this cleaning step to script file

https://github.com/pantheon-systems/pantheon-content-publisher-drupal/blob/42ba011ead94f7951d7210644850ba31055cc9e3/.github/scripts/cleanup-multidevs.sh

tundeafolabi commented on .github/workflows/pull_request.yml:
```diff
@@ -3,31 +3,147 @@ on:
+  workflow_dispatch:
+    inputs:
+      tmate_enabled:
+        description: Enable tmate debugging
```
for my knowledge what does tmate debugging enable here?

rkunjappan replied:
tmate is a fork of tmux that allows live SSH access to a GitHub Actions runner for debugging failing CI jobs. But by default, it's disabled here. We can enable it by running gh workflow run pull_request.yml --ref SITE-5259-unit-tests -f tmate_enabled=1. When the workflow hits the "Setup tmate session" step, it will pause and display an SSH string in the logs (e.g., ssh XYZ@nyc1.tmate.io). We can use this to connect to the GitHub runner shell and run commands like terminus or drush directly on the runner.

<img width="562" height="815" alt="Screenshot 2026-02-25 at 1 12 44 PM" src="https://github.com/user-attachments/assets/6f8d1b15-645a-48d6-b862-7e29ab0b405c" />

rkunjappan replied:
Some other modules were using tmate, but it’s an unnecessary step in this workflow (once Phil mentioned that). When I initially tried removing it, it caused some issues, so I kept it for stability. I will now try to remove the tmate script.

rkunjappan committed:
Fix  case sensitivity of publishStatus

AnaisPantheor reviewed (approved):
LGTM

rkunjappan committed:
Extract multidev cleanup logic into standalone script

Move inline cleanup bash from pull_request.yml into
.github/scripts/cleanup-multidevs.sh for consistency with
existing create-multidev.sh and run-tests.sh scripts.

rkunjappan committed:
Remove tmate debugging steps from pull_request workflow

tundeafolabi reviewed (approved):
LGTM
